### ISSUE TITLE
Automatically populating PMID & PMCID fields when generating CSL items for DOIs 

### ISSUE THREAD

**Issue by dhimmel** [greenelab/manubot#45]
Would it be nice for us to automatically fill in PMID and PMCID for DOI CSL items, when possible?

For example, we'd also probably change the default bibliographic style to be like:

> **Sci-Hub provides access to nearly all scholarly literature**
Daniel S Himmelstein, Ariel Rodriguez Romero, Jacob G Levernier, Thomas Anthony Munro, Stephen Reid McLaughlin, Bastian Greshake Tzovaras, Casey S Greene
_eLife_ (2018-03-01) https://doi.org/ckcj
DOI: 10.7554/elife.32822 · PMID: 29424689 · PMCID: PMC5832410

I am thinking this could be helpful for some journals which like to show all IDs when possible. It could also be generally useful and make readers more cognizant of IDs.

**Comment by dhimmel** [comment]
# PMC ID converter API

PubMed Central has an [ID converter API](https://www.ncbi.nlm.nih.gov/pmc/tools/id-converter-api/), but it only works for articles in PubMed Central and does not work for articles in PubMed but not PMC.

For example

```
https://www.ncbi.nlm.nih.gov/pmc/utils/idconv/v1.0/?tool=my_tool&email=my_email@example.com&ids=10.7554/elife.32822
```

returns

```xml
<pmcids status="ok">
  <request idtype="doi" dois="" versions="yes" showaiid="no">
    <echo>tool=my_tool;email=my_email%40example.com;ids=10.7554%2Felife.32822</echo>
  </request>
  <record requested-id="10.7554/ELIFE.32822" pmcid="PMC5832410" pmid="29424689" doi="10.7554/eLife.32822">
    <versions>
      <version pmcid="PMC5832410.1" current="true" />
    </versions>
  </record>
</pmcids>
```

**Comment by dhimmel** [comment]
## Entrez e-Utilities

Finding PMIDs for articles in PubMed but not PMC is going to be a bit harder. However, possibly we could use a query like:

```
https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&term=10.1161/CiRCGENETICS.115.001181[DOI]
```

```xml
<eSearchResult>
  <Count>1</Count>
  <RetMax>1</RetMax>
  <RetStart>0</RetStart>
  <IdList>
    <Id>27094199</Id>
  </IdList>
  <TranslationSet />
  <TranslationStack>
    <TermSet>
      <Term>10.1161/CiRCGENETICS.115.001181[DOI]</Term>
      <Field>DOI</Field>
      <Count>1</Count>
      <Explode>N</Explode>
    </TermSet>
    <OP>GROUP</OP>
  </TranslationStack>
  <QueryTranslation>10.1161/CiRCGENETICS.115.001181[DOI]</QueryTranslation>
</eSearchResult>
```

**Comment by dhimmel** [comment]
See PR https://github.com/greenelab/manubot/pull/46 where this has been implemented using the following strategy:

1. Try to use the PMC ID converter API
2. If not found in PMC, use NCBI's eSearch to get the PMID

Currently in that PR, this feature has been added as part of the DOI metadata pipeline. Perhaps it should be optional but enabled by default? It does increase the number of API calls per DOI from 2 to 3/4, and of course increases the possibility for things to go wrong. Although, I did try to implement it in a forgiving way (errors shouldn't generally crash the program).
